packages/react/src/hooks/useCallbackState.ts

```diff
@@ -0,0 +1,69 @@
+import { useCallback, useDebugValue, useMemo, useState } from 'react';
+
+type SetStateType<T> = T | ((prev: T) => T);
+type CallBackType<T> = (updatedValue: T) => void;
+
+/**
+ * A wrapper around the useState hook with an optional onChange callback.
+ *
+ * Mostly based on [1], but inspired by [2] to make the callback function a "dumb"
+ * callback and have no bearing on whether the invocation of setState actually
+ * updates the underlying state
+ *
+ * [1]. https://github.com/theKashey/use-callback-state
+ * [2]. https://medium.com/geekculture/usecallbackstate-the-hook-that-let-you-run-code-after-a-setstate-operation-finished-25f40db56661
+ * @param options - An object
+ * @param options.initialState - The value you want the state to be initially. It can be a value of any type, but there is a special behavior for functions. This argument is ignored after the initial render.
+ * @param options.debugMode - enable debug console messages
+ * @param options.debugName - name to be used in debug messages
+ * @returns [state, setState]
+ */
+function useCallbackState<T>({
```

Comment:
Is this primarily for dev and testing?

I didn't see any calls to this hook using `debugMode` or `debugName`.

Perf impact at runtime?

Reply:
So, the reason I had introduced `useCallbackState` is complicated to say the least.

We often follow the pattern of propagating the chain of `onChange` callbacks with something like this

```typescript
function setValueWrapper(newValue: any): void {
  setValue(newValue);
  if (props.onChange) {
    props.onChange(newValue);
  }
}

return (
...
<ChildInput onChange={setValueWrapper} />
...
);
```

In the before times, where every change in the form was triggered by the user value, this pattern works out just fine.

Now, consider the after times where fixed (and pattern) values need to be added in to the resource's value at various levels of the hierarchy. These fixed values are determined shortly after (importantly, a later render cycle) input components are rendered. A common scenario for fixed values are extension URLs. This is seemingly straightforward enough still when looking at an individual Extension, but what about an array of Extensions?

A `ResourceArrayInput` will have several `SliceInput`/`ExtensionInput` below it, all of which attempt to apply fixed values in the same render cycles. Finally we arrive at the issue: using the pattern above, a race condition exists since each call of `setValue` on the `ResourceArrayInput` would mask the previous invocations since they're working on a stale version of `value`, resulting in only the changes from whatever the last extension in the array happens to be persisting to the `ResourceArrayInput` value:

```
function setValuesWrapper(newValue: any, valueIndex: number) {
  const newValues = [...values];
  newValues[valueIndex] = newValue;
  setValues(newValues);
  if (props.onChange) {
    props.onChange(newValue);
  }
}
```

React has the masking aspect of this covered with the callback version of `setState`. However, a new problem arises: calling `onChange` in a setState callback as show below can lead to errors like the one shown below depending on the exact interplay of events, effects, and render cycles:

```typescript
function setValuesWrapper(newValue: any, valueIndex: number) {
  setValues((oldValues) {
    const newValues = [...oldValues];
    newValues[valueIndex] = newValue;

    if (props.onChange) {
      // Danger! Calling onChange in a setState callback can lead to React errors
      props.onChange(newValue);
    }
    return newValues;
  });
}
```

![Screenshot 2024-01-22 at 7 22 51 PM](https://github.com/medplum/medplum/assets/933303/1c00e7fb-0392-4069-b020-ebdab1e61ab0)

Error message from the screenshot above:

Cannot update a component (`BackboneElementInput`) while rendering a different component (`ElementsInput`). To locate the bad setState() call inside `ElementsInput`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render

Instead, a `useEffect` needs to be added to `values` for changes and only then calling `onChanges`. Putting all that together gets this most of the way there, but the onChange callback chain approach is definitely starting to feel brittle. As we've discussed, at some point I think it'd be worth investigating switching to an uncontrolled form approach instead.

Oh, also, `useCallbackState` turned out to not even be effective at what I thought it was doing. So that'll be going away.